geode-core/src/distributedTest/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceConstraintsDistributedTest.java

```diff
@@ -0,0 +1,595 @@
+        break;
+      case INVALIDATE:
+        assertThat(region.get(INVALIDATE_KEY)).isNull();
+        break;
+    }
+  }
+
+  private void executeOperationFromClient(String regionName, Operation operation,
+      QueryObject testValue) {
+    client.invoke(() -> {
+      assertThat(ClusterStartupRule.getClientCache()).isNotNull();
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      Region<Integer, QueryObject> region =
+          clientCache.<Integer, QueryObject>createClientRegionFactory(ClientRegionShortcut.PROXY)
+              .create(regionName);
+
+      ThrowableAssert.ThrowingCallable operationCallable =
+          getRegionOperation(operation, region, testValue);
+      assertThatCode(operationCallable).doesNotThrowAnyException();
+    });
+  }
+
+  /**
+   * The test registers a {@link QueryObserver} that changes the installed
+   * {@link MethodInvocationAuthorizer} to the default one (denies everything) as soon as the query
+   * starts, and makes sure the query succeeds (it is no affected by the authorizer change).
+   * The query is executed from the client.
+   */
+  @Test
+  @Parameters({"LOCAL", "REPLICATE", "PARTITION"})
+  @TestCaseName("[{index}] {method}(RegionType:{0})")
+  public void queriesInFlightExecutedByClientsShouldNotBeAffectedWhenMethodAuthorizerIsChanged(
+      RegionShortcut regionShortcut) {
+    String regionName = testName.getMethodName();
+    createAndPopulateRegion(regionName, regionShortcut);
+    String queryString = "<TRACE> SELECT object.getName() FROM /" + regionName + " object";
+
+    // Set test query observer.
+    server.invoke(() -> {
+      TestQueryObserver queryObserver =
+          new TestQueryObserver(RestrictedMethodAuthorizer.class.getName(), Collections.emptySet());
+      QueryObserverHolder.setInstance(queryObserver);
+    });
+
+    // Execute query from the client.
+    client.invoke(() -> {
+      assertThat(ClusterStartupRule.getClientCache()).isNotNull();
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      Query query = clientCache.getQueryService().newQuery(queryString);
+      @SuppressWarnings("unchecked")
+      SelectResults<QueryObject> result = (SelectResults<QueryObject>) query.execute();
+      assertThat(result.size()).isEqualTo(ENTRIES);
+    });
+
+    // Execute query again, it should fail as the restricted authorizer has been installed.
+    server.invoke(QueryObserverHolder::reset);
+    client.invoke(() -> {
+      assertThat(ClusterStartupRule.getClientCache()).isNotNull();
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      Query newQuery = clientCache.getQueryService().newQuery(queryString);
+      assertThatThrownBy(newQuery::execute)
+          .isInstanceOf(ServerOperationException.class)
+          .hasCauseInstanceOf(NotAuthorizedException.class)
+          .hasStackTraceContaining(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "getName");
```

Comment:
Here (and in other tests) could the "getName" (or "getId" if that's the method used) string be extracted to a variable so that we can avoid having the string hardcoded in two different places?

Reply:
Done!.